imagebuildah/stage_executor.go

```diff
@@ -1164,6 +1164,30 @@ func (s *StageExecutor) Execute(ctx context.Context, base string) (imgID string,
 		// determining if a cached layer with the same build args already exists
 		// and that is done in the if block below.
 		if checkForLayers && step.Command != "arg" && !(s.executor.squash && lastInstruction && lastStage) {
+			if step.Command == "add" || step.Command == "copy" {
+				// There is no way we are going to hit cache for `ADD`
+				// and `COPY` without adding content to find the digest
+				// of the content for which we need to check the cache for
+				// hence we can save cycles if we created `addedContentSummary`
+				// for these special cases by actually adding content so we
+				// end up hitting cache in the first pass.
+				if err = ib.Run(step, s, noRunsRemaining); err != nil {
+					logrus.Debugf("Error building at step %+v: %v", *step, err)
+					return "", nil, fmt.Errorf("error building at STEP \"%s\": %w", step.Message, err)
+				}
+
+				// In case we added content, retrieve its digest.
+				addedContentType, addedContentDigest := s.builder.ContentDigester.Digest()
+				addedContentSummary = addedContentType
+				if addedContentDigest != "" {
+					if addedContentSummary != "" {
+						addedContentSummary = addedContentSummary + ":"
+					}
+					addedContentSummary = addedContentSummary + addedContentDigest.Encoded()
+					logrus.Debugf("added content %s", addedContentSummary)
+				}
```

Comment:
At three copies, this should _definitely_ be turned into a function. (The 480-line `Execute` function is already so large, so I’m looking for ways to make it smaller.)